src/record_syscall.cc

```diff
@@ -3917,6 +3926,38 @@ static void process_execve(RecordTask* t, TaskSyscallState& syscall_state) {
 
   t->post_exec_syscall(*syscall_state.exec_saved_event);
 
+  // Check if the executable we just execed has setuid bits or file capabilities
+  // If so, they may have been ignored, due to our no_new_privs setting in the
+  // tracee. That's most likely not what the user intended (and setuid
+  // applications may not handle not being root particularly gracefully - after
+  // all under usual circumstances, it would be an exec-time error). Give a loud
+  // warning to tell the user what happened, but continue anyway.
+  {
+    // Only issue the warning once. If it's a problem, the user will likely
+    // find out soon enough. If not, no need to keep bothering them.
+    static bool gave_stern_warning = false;
+    if (!gave_stern_warning && !t->session().setuid_sudo_mode() && geteuid()) {
+      struct vfs_cap_data actual, empty;
+      memset(&(actual.data), 0, sizeof(actual.data));
+      memset(&(empty.data), 0, sizeof(empty.data));
+      if (0 == getxattr(t->vm()->exe_image().c_str(), "security.capability",
+                        &actual, sizeof(vfs_cap_data))) {
+        if (memcmp(&actual, &empty, sizeof(actual.data)) != 0) {
+          fputs(dropped_privs_warning, stderr);
+          gave_stern_warning = true;
+        }
+      } else {
+        ASSERT(t, errno == ENODATA);
+        struct stat buf;
+        stat(t->vm()->exe_image().c_str(), &buf);
+        if (buf.st_mode & (S_ISUID | S_ISGID)) {
+          fputs(dropped_privs_warning, stderr);
+          gave_stern_warning = true;
```

Comment:
Probably should avoid code duplication b y moving the "is setuid/setgid` into a helper function.